message_ix/core.py

```diff
@@ -0,0 +1,128 @@
+        """
+        jobj = platform._jobj
+        if version == 'new':
+            scheme = 'MESSAGE'
+            jscen = jobj.newScenario(model, scen, scheme, annotation)
+        elif isinstance(version, int):
+            jscen = jobj.getScenario(model, scen, version)
+        else:
+            jscen = jobj.getScenario(model, scen)
+
+        super(Scenario, self).__init__(
+            platform, model, scen, jscen, cache=cache)
+
+    def add_spatial_sets(self, data):
+        """Add sets related to spatial dimensions of the model
+
+        Parameters
+        ----------
+        data : dict or other
+
+        Examples
+        --------
+        data = {'country': 'Austria'}
+        data = {'country': ['Austria', 'Germany']}
+        data = {'country': {'Austria': {'state': ['Vienna', 'Lower Austria']}}}
+        """
+        nodes = []
+        levels = []
+        hierarchy = []
+
+        def recurse(k, v, parent='World'):
+            if isinstance(v, collections.Mapping):
+                for _parent, _data in v.items():
+                    for _k, _v in _data.items():
+                        recurse(_k, _v, parent=_parent)
+
+            level = k
+            children = [v] if isscalar(v) else v
+            for child in children:
+                hierarchy.append([level, child, parent])
+                nodes.append(child)
+            levels.append(level)
+
+        for k, v in data.items():
+            recurse(k, v)
+
+        self.add_set("node", nodes)
+        self.add_set("lvl_spatial", levels)
+        self.add_set("map_spatial_hierarchy", hierarchy)
+
+    def add_horizon(scenario, data):
+        """Add sets related to temporal dimensions of the model
+
+        Parameters
+        ----------
+        scenario : ixmp.Scenario
+        data : dict or other
+
+        Examples
+        --------
+        data = {'year': [2010, 2020]}
+        data = {'year': [2010, 2020], 'firstmodelyear': 2020}
+        """
+        if 'year' not in data:
+            raise ValueError('"year" must be in temporal sets')
+        horizon = data['year']
+        scenario.add_set("year", horizon)
+
+        first = data['firstmodelyear'] if 'firstmodelyear' in data else horizon[0]
+        scenario.add_cat("year", "firstmodelyear", first)
+
+    def add_temporal_sets(self, data):
+        """Add sets related to temporal dimensions of the model
+
+        Parameters
+        ----------
+        data : dict or other
+
+        Examples
+        --------
+        data = {'year': [2010, 2020]}
+        data = {'year': [2010, 2020], 'firstmodelyear': 2020}
+        """
+        if 'year' not in data:
+            raise ValueError('"year" must be in temporal sets')
+        horizon = data['year']
+        self.add_set("year", horizon)
+
+        first = data['firstmodelyear'] if 'firstmodelyear' in data else horizon[0]
+        self.add_set("cat_year", ["firstmodelyear", first])
```

Comment:
please change to `self.add_cat('year', 'firstmodelyear', first, is_unique=True)`